File: Cplusplus/LEVEL 2/행렬 테두리 회전하기.cpp
//22.02.01 행렬 테두리 회전하기 by CJHofficial

#include <string>
#include <vector>

using namespace std;

int arr[102][102] = { 0, };

vector<int> solution(int rows, int columns, vector<vector<int>> queries) {
    vector<int> answer;
    for (int i = 1; i <= rows; i++) {
        for (int j = 1; j <= columns; j++) {
            arr[i][j] = j + columns * (i - 1);
        }
    }
    int startRow, startCol, finalRow, finalCol, temp1, temp2;
    for (auto v : queries) {
        startRow = v[0]; startCol = v[1]; finalRow = v[2]; finalCol = v[3];
        temp1 = arr[startRow][startCol];
        int min = temp1;
        for (int j = startCol; j < finalCol; j++) {
            if (min > temp1) min = temp1;
            temp2 = arr[startRow][j + 1];
            arr[startRow][j + 1] = temp1;
            temp1 = temp2;
        }
        for (int i = startRow; i < finalRow; i++) {
            if (min > temp1) min = temp1;
            temp2 = arr[i + 1][finalCol];
            arr[i + 1][finalCol] = temp1;
            temp1 = temp2;
        }
        for (int j = finalCol; j > startCol; j--) {
            if (min > temp1) min = temp1;
            temp2 = arr[finalRow][j - 1];
            arr[finalRow][j - 1] = temp1;
            temp1 = temp2;
        }
        for (int i = finalRow; i > startRow; i--) {
            if (min > temp1) min = temp1;
            temp2 = arr[i - 1][startCol];
            arr[i - 1][startCol] = temp1;
            temp1 = temp2;
        }
        answer.push_back(min);
    }
    return answer;
}

File: Python/level2/행렬 테두리 회전하기.py
def solution(rows, columns, queries):
    answer = []
    number = 0
    array = [[row * columns + col for col in range(1, columns + 1)] for row in range(rows)]
    for query in queries:
        answer.append([])
        [leftX, leftY, rightX, rightY] = query
        [tempX, tempY] = [leftX, leftY]
        temp1 = array[tempX - 1][tempY - 1]
        answer[-1].append(temp1)
        for y in range(leftY, rightY):
            temp2 = array[leftX - 1][y]
            array[leftX - 1][y] = temp1
            temp1 = temp2
            answer[-1].append(temp1)
        for x in range(leftX, rightX):
            temp2 = array[x][rightY - 1]
            array[x][rightY - 1] = temp1
            temp1 = temp2
            answer[-1].append(temp1)
        for y in range(rightY - 2, leftY - 2, -1):
            temp2 = array[rightX - 1][y]
            array[rightX - 1][y] = temp1
            temp1 = temp2
            answer[-1].append(temp1)
        for x in range(rightX - 2, leftX - 2, -1):
            temp2 = array[x][leftY - 1]
            array[x][leftY - 1] = temp1
            temp1 = temp2
            answer[-1].append(temp1)
        answer[-1] = sorted(answer[-1])[0]
    return answer
